File: src/main/java/Sorting/InsertionSort.java
package main.java.Sorting;

public class InsertionSort {

    public static void iterative_sort(int[] arr) {

        if (arr.length <= 1)
            return;

        for (int i = 1; i < arr.length; i++) {

            int j = i - 1;
            int temp = arr[i];

            while (j >= 0 && arr[j] > temp) {

                arr[j + 1] = arr[j];
                j--;
            }

            arr[j + 1] = temp;
        }
    }

    private static void recursive_sort(int[] arr, int i) {

        if (i >= arr.length)
            return;

        int j = i - 1;
        int temp = arr[i];

        while (j >= 0 && arr[j] > temp) {

            arr[j + 1] = arr[j];
            j--;
        }

        arr[j + 1] = temp;

        recursive_sort(arr, i + 1);
    }

    public static void recursive_sort(int[] arr) {

        if (arr.length <= 1)
            return;

        recursive_sort(arr, 1);
    }
}

File: src/main/python/Sorting/insertion_sort.py
class InsertionSort:

    @staticmethod
    def iterative_sort(arr: list[int]) -> None:
        if len(arr) <= 1:
            return

        for i in range(1, len(arr)):
            j = i - 1
            temp = arr[i]

            while j >= 0 and arr[j] > temp:
                arr[j + 1] = arr[j]
                j -= 1

            arr[j + 1] = temp

    @staticmethod
    def recursive_func(arr: list[int], i: int) -> None:
        if i >= len(arr):
            return

        j = i - 1
        temp = arr[i]

        while j >= 0 and arr[j] > temp:
            arr[j + 1] = arr[j]
            j -= 1

        arr[j + 1] = temp

        InsertionSort.recursive_func(arr, i + 1)

    @staticmethod
    def recursive_sort(arr: list[int]) -> None:
        if len(arr) <= 1:
            return

        InsertionSort.recursive_func(arr, 1)